nifi-commons/nifi-questdb/src/test/java/org/apache/nifi/questdb/CompositeClientTest.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.questdb;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.junit.jupiter.MockitoSettings;
+import org.mockito.quality.Strictness;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.BiConsumer;
+
+import static org.apache.nifi.questdb.QuestDbTestUtil.SELECT_QUERY;
+
+// This test uses the {@code compile} method to cover different scenarios. It is expected that other methods behave the same
+// but the coverage of the individual behaviour of the individual methods is provided by different tests.
+@ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
+public class CompositeClientTest extends EmbeddedQuestDbTest {
+    private static final int NUMBER_OF_RETRIES = 3;
+    private static final int NUMBER_OF_ATTEMPTS = NUMBER_OF_RETRIES + 1;
+    private static final int LOCK_ATTEMPT_TIME = 20;
+    private static final TimeUnit LOCK_ATTEMPT_TIME_UNIT = TimeUnit.MILLISECONDS;
+
+    @Mock
+    Client client;
+
+    @Mock
+    Client fallback;
+
+    @Mock
+    BiConsumer<Integer, Exception> errorAction;
+
+    @Mock
+    ConditionAwareClient.Condition condition;
+
+    private ReadWriteLock databaseStructureLock;
+
+    private LockedClient lockedClientSpy;
+
+    private Client testSubject;
+
+    @BeforeEach
+    public void setUp() {
+        databaseStructureLock = new ReentrantReadWriteLock();
+
+        testSubject = getTestSubject();
+
+        Mockito.when(condition.check()).thenReturn(true);
```

Comment:
I recommend using static imports for Mockito and Assertion methods for more concise test code.